server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java

```diff
@@ -183,20 +184,23 @@ private void disconnectClient(ChannelHandlerContext ctx) {
     }
 
     @Override
-    public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) throws Exception {
+    public void messageReceived(ChannelHandlerContext ctx, MessageEvent event) throws Exception {
         try (Closeable closeable = IMAPMDCContext.from(ctx, attributes)) {
             imapCommandsMetric.increment();
             ImapSession session = (ImapSession) attributes.get(ctx.getChannel());
             ImapResponseComposer response = (ImapResponseComposer) ctx.getAttachment();
-            ImapMessage message = (ImapMessage) e.getMessage();
+            ImapMessage message = (ImapMessage) event.getMessage();
             ChannelPipeline cp = ctx.getPipeline();
 
             try {
-                if (cp.get(NettyConstants.EXECUTION_HANDLER) != null) {
-                    cp.addBefore(NettyConstants.EXECUTION_HANDLER, NettyConstants.HEARTBEAT_HANDLER, heartbeatHandler);
-                } else {
-                    cp.addBefore(NettyConstants.CORE_HANDLER, NettyConstants.HEARTBEAT_HANDLER, heartbeatHandler);
-
+                try {
+                    if (cp.get(NettyConstants.EXECUTION_HANDLER) != null) {
+                        cp.addBefore(NettyConstants.EXECUTION_HANDLER, NettyConstants.HEARTBEAT_HANDLER, heartbeatHandler);
+                    } else {
+                        cp.addBefore(NettyConstants.CORE_HANDLER, NettyConstants.HEARTBEAT_HANDLER, heartbeatHandler);
+                    }
+                } catch (IllegalArgumentException e) {
+                    LOGGER.info("heartbeat handler is already part of this pipeline", e);

```

Comment:
Is this nested try/catch block necessary? Could we bring this outside?

Reply:
No because I want to proceed with the request.

Moving it outside will result in a different behaviour.